### DISCUSSION TITLE
Loading self compiled emqx_auth_http plugin

### DISCUSSION THREAD

**Discussion by sushantAtAws** [emqx/emqx#5203] [Q&A] [answered]
Hi compiled emqx and emqx_auth_http from scratch. Started emqx and copied over the plugin to the lib directory. Now I am trying to load the plugin but it fails to locate it.

`cp -r /home/ubuntu/emqx-auth-http/_build/default/lib/emqx_auth_http /home/ubuntu/emqx/_build/emqx/rel/emqx/lib`

```
./bin/emqx_ctl plugins load emqx_auth_http
Load plugin emqx_auth_http error: not_found.
```

1. Is this method not supported?
2. How am I suppose to load this compiled plugin? My next step is to customize the http auth plugin and load that.
3. How is the released version of emqx packaged? Is that in a separate repo?

**Comment by zmstone** [comment]
Hi 
If you do not want to clone  the repo and make the changes, but compile the plugin independently.
Please follow the steps here https://www.emqx.io/blog/develop-and-deploy-emqx-plugin-for-enterprise-4-3
It’s for enterprise, but should work for open source too.

**Reply by sushantAtAws** [reply]
Hi, thanks @zmstone 
I tried the instructions you shared.

Based on my config, the plugins should live in the `plugins` folder:
```
cat etc/emqx.conf | grep expand_plugins_dir
    expand_plugins_dir: "plugins/"
```
So I copied the plugin artifacts to the `plugin` folder and restarted the server.
```
ubuntu@ip-172-31-53-162:~/emqx/_build/emqx/rel/emqx/plugins$ ls
emqx_auth_http  emqx_reloader  ssl_verify_fun
```

On loading the plugin, I get the following error:
`./bin/emqx_ctl plugins load emqx_auth_http`:

```
2021-07-13T19:47:15.562938+00:00 [error] crasher: initial call: application_master:init/4, pid: <0.1860.0>, registered_name: [], exit: {{different_server,{emqx_auth_http_app,start,[normal,[]]}},[{application_master,init,4,[{file,"application_master.erl"},{line,138}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}, ancestors: [<0.1859.0>], message_queue_len: 1, messages: [{'EXIT',<0.1861.0>,normal}], links: [<0.1859.0>,<0.1215.0>], dictionary: [], trap_exit: true, status: running, heap_size: 610, stack_size: 27, reductions: 238; neighbours:
2021-07-13T19:47:15.563376+00:00 [error] [Plugins] Load plugin emqx_auth_http failed, cannot start plugin emqx_auth_http for {different_server,{emqx_auth_http_app,start,[normal,[]]}}
```

Is there a version mismatch happening here?
I have compiled both the emqx server and the auth plugin from source with instructions:

```
git clone https://github.com/emqx/emqx.git
cd emqx
git checkout $(git describe --abbrev=0 --tags)
make
```
```
git clone https://github.com/emqx/emqx-auth-http.git
cd emqx-auth-http
git checkout $(git describe --abbrev=0 --tags)
make
```

**Reply by zmstone** [reply]
I’m afraid you need to pick a different name for your plugin.

**Reply by sushantAtAws** [reply]
Thanks. I will give that a try.

**Reply by sushantAtAws** [reply]
Hi @zmstone 
I compiled the plugin with different name, copied over and tried loading. Got the same error.

```
ubuntu@ip-172-31-53-162:~/emqx/_build/emqx/rel/emqx$ ./bin/emqx_ctl plugins load emqx_auth_http_2
Load plugin emqx_auth_http_2 error: {emqx_auth_http_2,
                                     {different_server,
                                      {emqx_auth_http_2_app,start,
                                       [normal,[]]}}}.
```

broker logs:
```
(emqx@127.0.0.1)1> 2021-07-13T22:55:30.191103+00:00 [error] crasher: initial call: application_master:init/4, pid: <0.1817.0>, registered_name: [], exit: {{different_server,{emqx_auth_http_2_app,start,[normal,[]]}},[{application_master,init,4,[{file,"application_master.erl"},{line,138}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}, ancestors: [<0.1816.0>], message_queue_len: 1, messages: [{'EXIT',<0.1818.0>,normal}], links: [<0.1816.0>,<0.1215.0>], dictionary: [], trap_exit: true, status: running, heap_size: 610, stack_size: 27, reductions: 251; neighbours:
2021-07-13T22:55:30.191592+00:00 [error] [Plugins] Load plugin emqx_auth_http_2 failed, cannot start plugin emqx_auth_http_2 for {different_server,{emqx_auth_http_2_app,start,[normal,[]]}}
```

Is there any other log I should be looking at to get more debug info?

**Comment by zmstone** [comment, marked as answer]
sorry I am on my phone, did not notice the branches you were using 

the docs i shared were tested in v4.3, so 

1. Not related to the issue, but highly recommended: Please checkout `main-v4.3` branch from emqx.git
The master branch is not quite stable yet.

2.
modify apps/emqx_auth_http per your requirements, or copy apps/emqx_auth_http to a independent rebar3 project 
emqx_auth_http_2 then comple & release by following the steps in the document.
NOTE: all occurrences of  emqx_auth_http should be replaced by emqx_auth_http_2 including config file.

3.
different_server seem to be this error:
https://github.com/emqx/emqx-auth-http/blob/master/src/emqx_auth_http_app.erl#L156
The plugin doesn’t allow urls in config to differ in host name for auth_req, acl_req etc.
Please make your changes based on apps/emqx_auth_http which made some changes in the config layout

**Reply by sushantAtAws** [reply]
Thanks for taking a look.
I will give this a shot and update.

**Reply by sushantAtAws** [reply]
Hi @zmstone 

I was able to get it to work. Thanks to you.

Want to mention the steps I took:
1. I had build failures with the head of `main-v4.3` branch. So I reset to the a known verified commit `11352c4eaae76fec974776129f20eb69f8bcc528`. With this I was able to build 4.3 release and start it.
2. For the emqx_auth_http plugin:
2.1 I cloned the repo and went to 4.2.11
    ```
    git clone https://github.com/emqx/emqx-auth-http.git
    cd emqx-auth-http
    git checkout $(git describe --abbrev=0 --tags)
    ```
    2.2 Copy over the code from emqx's apps/emqx_auth_http directory
    2.3 Change all the references from `emqx_auth_http` to `emqx_auth_http_3` 
    2.4 Built with `make`
3. Copied over the artifacts: 
`cp -r emqx_auth_http_3 ~/emqx3/emqx/_build/emqx/rel/emqx/etc/plugins/`
3.1 One missing piece from the documentation was that copying over the `ebin` was not enough. We need the config and schema files. Also need to fix the symlinks.
```
rm ~/emqx3/emqx/_build/emqx/rel/emqx/etc/plugins/emqx_auth_http_3/include ~/emqx3/emqx/_build/emqx/rel/emqx/etc/plugins/emqx_auth_http_3/priv ~/emqx3/emqx/_build/emqx/rel/emqx/etc/plugins/emqx_auth_http_3/src
mkdir ~/emqx3/emqx/_build/emqx/rel/emqx/etc/plugins/emqx_auth_http_3/etc
cp ~/emqx-auth-http-3/emqx-auth-http/etc/emqx_auth_http_3.conf ~/emqx3/emqx/_build/emqx/rel/emqx/etc/plugins/emqx_auth_http_3/etc/
mkdir ~/emqx3/emqx/_build/emqx/rel/emqx/etc/plugins/emqx_auth_http_3/priv
cp ~/emqx-auth-http-3/emqx-auth-http/priv/emqx_auth_http_3.schema ~/emqx3/emqx/_build/emqx/rel/emqx/etc/plugins/emqx_auth_http_3/priv/
```
4. Restarting the broker and loading the emqx_auth_http_3 plugin works after this.

I really appreciate all the help.

**Reply by zmstone** [reply]
great

> src and include are optional since all they have inside are source code. ebin directory is where the compiled beam files reside. etc is per EMQ X’s convention, the directory to store the plugin’s config file, and priv directory is, per Erlang’s convention, to keep build artifacts to be released (in our case, the plugin’s config schema file).

I think it’s pretty clear what to copy as quoted above, or is it some other place which is only mentioned ebony?
I’d like to know so we can fix it.

**Reply by zmstone** [reply]
> I had build failures with the head of main-v4.3 branch. So I reset to the a known verified commit 11352c4eaae76fec974776129f20eb69f8bcc528. With this I was able to build 4.3 release and start it.

Could you share more info about the build failure?

In fact, if you release your  plugin in this way, you don’t have to compile emqx, the plugin can be released to emqx installed from a built package, e.g zip, rpm, or deb

**Reply by sushantAtAws** [reply]
> great
> 
> > src and include are optional since all they have inside are source code. ebin directory is where the compiled beam files reside. etc is per EMQ X’s convention, the directory to store the plugin’s config file, and priv directory is, per Erlang’s convention, to keep build artifacts to be released (in our case, the plugin’s config schema file).
> 
> I think it’s pretty clear what to copy as quoted above, or is it some other place which is only mentioned ebony?
> I’d like to know so we can fix it.

You are correct. I think I missed part of it.

**Reply by sushantAtAws** [reply]
> > I had build failures with the head of main-v4.3 branch. So I reset to the a known verified commit [11352c4](https://github.com/emqx/emqx/commit/11352c4eaae76fec974776129f20eb69f8bcc528). With this I was able to build 4.3 release and start it.
> 
> Could you share more info about the build failure?
> 
I could not reproduce the build failure now with the latest from the branch.

> In fact, if you release your plugin in this way, you don’t have to compile emqx, the plugin can be released to emqx installed from a built package, e.g zip, rpm, or deb

That makes sense. Thanks.
